src/simulation/LoadGenerator.cpp

```diff
@@ -866,11 +1025,224 @@ LoadGenerator::manageOfferTransaction(
+    // generate siginificant load
+    uint64_t const baseInstructionCount = 3'000'000;
+    uint64_t const instructionsPerGuestCycle = 120;
+    uint64_t const instructionsPerHostCycle = 2355;
+
+    // Pick random number of cycles between bounds, respecting network limits
+    uint64_t maxInstructions =
+        networkCfg.mTxMaxInstructions - baseInstructionCount;
+
+    uint64_t guestCyclesMax = maxInstructions / instructionsPerGuestCycle;
+    guestCyclesMax = std::min(cfg.guestCyclesHigh, guestCyclesMax);
+    uint64_t guestCyclesMin = std::min(cfg.guestCyclesLow, guestCyclesMax);
+    uint64_t guestCycles =
+        rand_uniform<uint64_t>(guestCyclesMin, guestCyclesMax);
+
+    maxInstructions -= guestCycles * instructionsPerGuestCycle;
+    uint64_t hostCyclesMax = maxInstructions / instructionsPerHostCycle;
+    hostCyclesMax = std::min(cfg.hostCyclesHigh, hostCyclesMax);
+    uint64_t hostCyclesMin = std::min(cfg.hostCyclesLow, hostCyclesMax);
+    uint64_t hostCycles = rand_uniform<uint64_t>(hostCyclesMin, hostCyclesMax);
+
+    SorobanResources resources;
+    resources.footprint.readOnly = instance.readOnlyKeys;
+
+    // Must always read wasm and instance
+    releaseAssert(networkCfg.mTxMaxReadLedgerEntries > 1);
+    auto maxEntries = networkCfg.mTxMaxReadLedgerEntries - 2;
+    maxEntries = std::min(maxEntries, cfg.nDataEntriesHigh);
+    auto minEntries = std::min(maxEntries, cfg.nDataEntriesLow);
+    auto numEntries = rand_uniform<uint32_t>(minEntries, maxEntries);
+    for (uint32_t i = 0; i < numEntries; ++i)
+    {
+        auto lk = contractDataKey(instance.contractID, makeU32(i),
+                                  ContractDataDurability::PERSISTENT);
+        resources.footprint.readWrite.emplace_back(lk);
+    }
+
+    // Use 2'000 as an estimate for Wasm size
+    uint32_t maxKiloBytesPerEntry =
+        (networkCfg.mTxMaxReadBytes - 2'000) / numEntries / 1024;
+    maxKiloBytesPerEntry =
+        std::min(maxKiloBytesPerEntry, cfg.kiloBytesPerDataEntryHigh);
+    uint32_t minKiloBytesPerEntry =
+        std::min(maxKiloBytesPerEntry, cfg.kiloBytesPerDataEntryLow);
+    auto kiloBytesPerEntry =
+        rand_uniform<uint32_t>(minKiloBytesPerEntry, maxKiloBytesPerEntry);
+
+    auto guestCyclesU64 = makeU64(guestCycles);
+    auto hostCyclesU64 = makeU64(hostCycles);
+    auto numEntriesU32 = makeU32(numEntries);
+    auto kiloBytesPerEntryU32 = makeU32(kiloBytesPerEntry);
+
+    Operation op;
+    op.body.type(INVOKE_HOST_FUNCTION);
+    auto& ihf = op.body.invokeHostFunctionOp().hostFunction;
+    ihf.type(HOST_FUNCTION_TYPE_INVOKE_CONTRACT);
+    ihf.invokeContract().contractAddress = instance.contractID;
+    ihf.invokeContract().functionName = "do_work";
+    ihf.invokeContract().args = {guestCyclesU64, hostCyclesU64, numEntriesU32,
+                                 kiloBytesPerEntryU32};
+
+    // We don't have a good way of knowing how many bytes we will have to read,
```

Comment:
That's a bit surprising, is there no way to come up with some upper bound on the read size?

Reply:
Well the upper bound is easy, it's the TX readBytes limit since that's what we base all our random values on. The issue is the current read values depend on the last invocation's write values, which are randomly selected. I thought about having some mapping of instance -> last invocation writeBytes, but the issue is there's no good way to maintain that mapping. We could update the writeBytes value when we submit the TX, but the TX may fail, leaving us with incorrect state. We'd have to check that the TX actually got applied to know the correct bytes value, and this would require that loadgen do disk lookups. Loadgen is already expensive as it is, so I think leaving these values as network max is fine.

Reply:
Ok, this probably is good enough initially. We could be more precise though if we set `kiloBytesPerDataEntryHigh * nDataEntriesHigh + <Wasm+instance size>` as the limit (this way it's still possible to have load that is not bounded by the read bytes when we skip the IO).
Going forward, we probably could come up with a way to improve this, but we definitely should get something working first.

Reply:
Can we update this as suggested above?

Reply:
Sorry, missed this comment.